core: dispatch Asset events

ndr-brt opened the pull request:
## What this PR changes/adds

Introduce events for domain class `Asset` and publish them through the `EventRouter`

## Why it does that

To permit asynchronous notifications

## Further notes

- Add a first documentation markdown about how to subscribe to and how to publish events
- Implement a default behavior for `AssetLoader.accept(Asset asset, DataAddress dataAddress)` to remove some duplication
- Add the event publishing in the `AssetService` class. Currently not every component that act on Assets pass through the service (e.g. `DataLoader`). So here I have a question: should `AssetService` moved to an upstream module (like the core module or a dedicated core-asset module?)
- Add serialization capabilities to all the events. The deserialization could be achived registering the event types on the `TypeManager`. This process is described in the doc markdown but since it's not needed (the EDC doesn't need to deserialize events for the moment) I left this out.
- The `EventRouter` publishes events in an asynchronous manner to avoid blocking the main thread and catches exceptions that could be thrown by the subscribers. It doesn't implement any retry behavior nor it uses a configurable `ExecutorService`. Those two functionalities could be addressed to further PRs.

## Linked Issue(s)

Closes #1435

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

paullatzelsperger commented on extensions/api/data-management/asset/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/asset/service/AssetServiceImpl.java:
```diff
@@ -56,6 +58,7 @@ public ServiceResult<Asset> create(Asset asset, DataAddress dataAddress) {
         return transactionContext.execute(() -> {
             if (findById(asset.getId()) == null) {
                 loader.accept(asset, dataAddress);
+                observable.invokeForEach(l -> l.created(asset));
```
maybe add a test to verify this. same for line 83

paullatzelsperger commented on extensions/api/data-management/asset/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/asset/service/AssetEventDispatchTest.java:
```diff
@@ -0,0 +1,72 @@
+
+package org.eclipse.dataspaceconnector.api.datamanagement.asset.service;
+
+import org.eclipse.dataspaceconnector.junit.launcher.EdcExtension;
```
I assume this was a rebase/merge artifact:
```suggestion
import org.eclipse.dataspaceconnector.junit.extensions.EdcExtension;
```

paullatzelsperger commented on spi/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/event/AssetDeleted.java:
```diff
@@ -0,0 +1,52 @@
+
+public class AssetDeleted extends Event {
+
+    private String id;
```
is this supposed to be the `assetId`? if so, i'd rename it to avoid ambivalence: one could be lead to believe this were the event's ID.

paullatzelsperger commented on spi/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/event/AssetDeleted.java:
```diff
@@ -0,0 +1,52 @@
+
+import java.util.Objects;
+
+public class AssetDeleted extends Event {
```
maybe a line of documentation?

ndr-brt replied:
@ndr-brt the name of the class is pretty self-explanatory, I find it a bit difficult to write something that doesn't sound totally useless :)

jimmarino replied:
How about what "deleted" means and what it entails, e.g. is "deletion" a marker and the Asset entity still in the backing store?

ndr-brt replied:
@jimmarino ok, got it, I would avoid the detail about the logical/phisical deletion, as it is an implementation delay of the store layer (I guess). 
I reported that, after the event emission, the asset won't be available anymore.

paullatzelsperger commented on spi/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/event/AssetCreated.java:
```diff
@@ -0,0 +1,52 @@
+
+import java.util.Objects;
+
+public class AssetCreated extends Event {
```
maybe a line of documentation?

paullatzelsperger commented on spi/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/event/Event.java:
```diff
@@ -0,0 +1,53 @@
+ * When serialized to JSON, will add a "type" field with the name of the runtime class name.
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type", include = JsonTypeInfo.As.EXTERNAL_PROPERTY)
+public abstract class Event {
```
Do you think it would be worthwhile to have the `at` field initialized with the current time by default? That would reduce the amount of duplicated code greatly, because then event sources wouldn't have to call `at(clock.millis())`.  
The general use of `Clock` makes this more cumbersome, but we could create a factory method of sorts, just to make it a bit sleeker to use, wdyt?

ndr-brt replied:
@paullatzelsperger this could go in the discussion about the event envelope mentioned in a comment above, if we do that we could create the event instance separately and then having someone (also the event router`) that would wrap the event with an envelope with other informations like `at`, an id, the type, and so on.

jimmarino replied:
> @jimmarino @paullatzelsperger , I added a commit where I introduced the `EventEnvelope` class, that contains the event and the metadata about the event (for the moment only the timestamp, but other information could be included in the future, as an id, the connector id and so on). I'm in doubt about the naming, because I'm not sure that the `Event` class now represent an event, since it misses the timestamp, an idea could be to rename `EventEnvelope` to `Event` and `Event` to `EventData` or something like this. wdyt?

Agreed the "Event" class should be the top-level name. Why not collapse the two classes into a single Event class? Generally, envelopes are introduced so that code can either process parts of a message without having to access the body (e.g. encrypted data) and/or to "unwrap" a message so that header information does not pollute non-transport areas of a code base.

BTW, this is a great PR! :-)

ndr-brt replied:
@jimmarino so your advice would be to have everything (data and metadata) inside the `Event` class as it was before my last commit?

paullatzelsperger commented on spi/core-spi/src/test/java/org/eclipse/dataspaceconnector/spi/event/AssetCreatedTest.java:
```diff
@@ -0,0 +1,45 @@
+                .build();
+
+        var json = typeManager.writeValueAsString(event);
+        System.out.println(json);
```
not std_out in tests pls (i assume you just forgot to remove it).

ndr-brt commented on extensions/api/data-management/asset/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/asset/service/AssetListener.java:
```diff
@@ -0,0 +1,44 @@
+     *
+     * @param asset the asset that has been created.
+     */
+    default void created(Asset asset) {
```
@paullatzelsperger not a big fan too. Will open a separate issue to avoid this also in other places

jimmarino replied:
FWIW, I like them when used to avoid boilerplate code :-)

ndr-brt commented on core/base/src/main/java/org/eclipse/dataspaceconnector/core/event/EventRouterImpl.java:
```diff
@@ -0,0 +1,54 @@
+                    if (throwable != null) {
+                        var subscriberName = subscriber.getClass().getSimpleName();
+                        var eventName = event.getClass().getSimpleName();
+                        monitor.severe(format("Subscriber %s failed to handle event %s", subscriberName, eventName), throwable);
```
@paullatzelsperger `RetryPolicy` and the `Result` are already on my list, will open an issue later on.
About the DLQ I'm not sure should be an edc-core responsability, I would leave this up to the extensions but we can talk about this.

jimmarino replied:
I think events should be fire-and-forget and non-reliable (i.e., no delivery guarantees). Otherwise, we will wind up building a reliable messaging platform. Another approach would be to architect reliability points like we do with the state machine implementations for specific cases. 

I don't think the events in this PR (e.g., asset created, etc.) need to be reliable. If someone needs this behavior, they should account for it in the client that issues the command that triggered the event.

 Let's discuss this further after this PR.

ndr-brt commented on spi/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/event/AssetCreated.java:
```diff
@@ -0,0 +1,52 @@
+
+public class AssetCreated extends Event {
+
+    private String id;
```
@paullatzelsperger You're right, I was thinking about the `Event` structure, maybe we could keep an `EventEnvelope` that contains an Event field so that we won't mind about field name conflicts. wdyt about this?

jimmarino commented on extensions/api/data-management/asset/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/asset/service/EventAssetListener.java:
```diff
@@ -0,0 +1,52 @@
+import org.eclipse.dataspaceconnector.spi.types.domain.asset.Asset;
+
+import java.time.Clock;
+
```
How about some doc here since it is in api?

ndr-brt commented:
@jimmarino @paullatzelsperger , I added a commit where I introduced the `EventEnvelope` class, that contains the event and the metadata about the event (for the moment only the timestamp, but other information could be included in the future, as an id, the connector id and so on).
I'm in doubt about the naming, because I'm not sure that the `Event` class now represent an event, since it misses the timestamp, an idea could be to rename `EventEnvelope` to `Event` and `Event` to `EventData` or something like this. wdyt?

ndr-brt commented on extensions/dataloading/src/main/java/org/eclipse/dataspaceconnector/dataloading/AssetLoader.java:
```diff
@@ -20,7 +20,9 @@
 
 public interface AssetLoader extends DataSink<AssetEntry> {
 
```
Is out of the scope of the PR, since I just refactored it. Will open a dedicated issue

ndr-brt replied:
https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/1467

ndr-brt commented:
@jimmarino @paullatzelsperger as we agreed I came back to the `Event` version without envelope, let's see how this will evolve with the next PRs
